src/v/raft/consensus.cc

```diff
@@ -1988,11 +1988,55 @@ consensus::do_append_entries(append_entries_request&& r) {
         co_return reply;
     }
 
-    // special case heartbeat case
+    model::offset batch_prev_log_index = request_metadata.prev_log_index;
+    if (batch_prev_log_index < last_log_offset) {
+        // The append point is before the end of our log. We need to skip
+        // over batches that we already have (they will have the matching
+        // term) to find the true truncation point. This is important for the
+        // case when we already have _all_ batches locally (possible if e.g.
+        // the request was delayed/duplicated). In this case we don't want to
+        // truncate, otherwise we might lose already committed data.
+
+        struct find_mismatch_consumer {
+            ss::future<ss::stop_iteration>
+            operator()(const model::record_batch& b) {
+                model::offset last_offset = last_matched
```

Comment:
We have this method in record_batch()

```
model::offset last_offset() const { return _header.last_offset(); 
```

I"m curious if we can use that instead of computing it.